core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1066,6 +1066,7 @@ class KafkaController(val config: KafkaConfig,
         // do this check only if the broker is live and there are no partitions being reassigned currently
         // and preferred replica election is not in progress
         val candidatePartitions = topicsNotInPreferredReplica.keys.filter(tp => controllerContext.isReplicaOnline(leaderBroker, tp) &&
+          controllerContext.partitionLeadershipInfo(tp).leaderAndIsr.isr.contains(leaderBroker) &&
```

Comment:
I think we should do this check last. We also want to use an option to avoid any potential NPEs. e.g:
```
controllerContext.partitionLeadershipInfo.get(partition).forall(...)
```

Reply:
Done.